src/Microsoft.Health.Dicom.Core/Extensions/DicomDatasetExtensions.cs

```diff
@@ -158,5 +188,57 @@ public static void AddValueIfNotNull<T>(this DicomDataset dicomDataset, DicomTag
                 dicomDataset.Add(dicomTag, value);
             }
         }
+
+        /// <summary>
+        /// Get DicomTags for index Tag from Dicom Dataset.
+        /// </summary>
+        /// <remarks>If indextag not exist in dataset, should not return.</remarks>
+        /// <param name="dicomDataset">The dicom dataset.</param>
+        /// <param name="indexTags">The index Tags.</param>
+        /// <returns>Mapping between IndexTag and DicomTag.</returns>
+        public static IDictionary<IndexTag, DicomTag> GetDicomTags(this DicomDataset dicomDataset, IEnumerable<IndexTag> indexTags)
+        {
```

Comment:
I am not sure about this whole method. Should'nt DicomTag be contained in IndexTag? Why are transforming it again based on the DicomDataset?

Reply:
Rename it to GetMatchingDicomTags which get matching dicom tag from dataset.
It's mostly for private tag -- we don't save private creator, so DicomDataset.GetSingleValueOrDefault(DicomTag) doesn't work for private tag, need to loop dataset and locate it by path.

Reply:
Can you add to the summary that this needs to exist for private tag?

---
In reply to: [590593679](https://github.com/microsoft/dicom-server/pull/612#discussion_r590593679) [](ancestors = 590593679)

Reply:
will put it in remark section

Reply:
I still don't get this. Are we missing a peace of information we need to add to PrivateTags then as part of AddTag?

Reply:
Yes. We don't have privateCreator for private tags.

Reply:
Will it help if we get this instead

Reply:
> Will it help if we get this instead
Hi Smitha,
Sorry, not get you.
We need DicomTag with privateCreator to call DicomDataSet.GetSingleOrDefaultValue(DicomTag) to retrieve value. Since we don't require customer to input privateCreator when create customtag, we have to loop through dataset to locate by tag path (groupid + elementid).

Reply:
These code will be cleaned up by https://microsofthealth.visualstudio.com/Health/_workitems/edit/80419